src/IonBinaryWriter.ts

```diff
@@ -654,13 +654,17 @@ class IntNode extends LeafNode {
   constructor(writer: LowLevelBinaryWriter, parent: Node, annotations: Uint8Array, private readonly value: number | JSBI) {
     super(writer, parent, IonTypes.INT, annotations);
 
-    if (this.value > 0) {
+    if (!(typeof this.value === 'number' || this.value instanceof JSBI)) {
+      throw new Error('Expected ' + this.value + ' to be a number or JSBI');
+    }
+
+    if (JSBI.GT(this.value, 0)) {
```

Comment:
This works fine, but:
* Using `JSBI.GT` will construct `JSBI` values from `number` values that it wishes to compare. That's a pretty high overhead for a simple `>` or `<` test.
* You can pass the static value `JsbiSupport.ZERO` instead of `0` to avoid instantiating `JSBI(0)` each time.

How about lifting the comparisons out?
```typescript
let isGreaterThanZero: boolean;
let isLessThanZero: boolean;
if (typeof value === 'number') {
   isGreaterThanZero = value > 0;
   isLessThanZero = value < 0;
} else if(value instanceof JSBI) {
   isGreaterThanZero = JsbiSupport.greaterThan(value, JsbiSupport.ZERO);
   isLessThanZero = JsbiSupport.greaterThan(value, JsbiSupport.ZERO);
} else {
  throw new Error('Expected ' + this.value + ' to be a number or JSBI');
}

if (isGreaterThanZero) {
  //...
} else if (isLessThanZero) {
  //...
} else {
  //...
}
```
I realize this is a larger restructuring than we had hoped for, but it's very explicit and avoids unnecessary overhead. It also folds the `assert` into the type checking.

Reply:
> Using JSBI.GT will construct JSBI values from number values that it wishes to compare.

@wesboyt made this observation the other day as well.  However, the logic of [JSBI.__compare()](https://github.com/GoogleChromeLabs/jsbi/blob/master/jsbi.mjs#L1077) indicates that this is not the case.  The JSBI source we depend on is a bit different, but doesn't appear to construct new JSBIs for ease-of-comparison either.

A case could be made that our code can perform less type-checking logic than JSBI internals, but I'm not sure that's warranted here (esp. in the absence of performance data).  If we're going to make such changes, let's make sure we're clear about *why* we're doing so.

Based on these findings, my inclination is to not lift the comparisons out as you've proposed.  Let me know if you still feel doing so would add clarity and/or material performance improvements.

Reply:
> However, the logic of [JSBI.__compare()](https://github.com/GoogleChromeLabs/jsbi/blob/master/jsbi.mjs#L1077) indicates that this is not the case. The JSBI source we depend on is a bit different, but doesn't appear to construct new JSBIs for ease-of-comparison either.

I stand corrected! It looks like the `(number, number)` case boils down to a number of type checks followed by [the expected operators](https://github.com/GoogleChromeLabs/jsbi/blob/master/jsbi.mjs#L1114-L1119).

> A case could be made that our code can perform less type-checking logic than JSBI internals, but I'm not sure that's warranted here (esp. in the absence of performance data). If we're going to make such changes, let's make sure we're clear about why we're doing so.

Agreed.

> Based on these findings, my inclination is to not lift the comparisons out as you've proposed. Let me know if you still feel doing so would add clarity and/or material performance improvements.

Your current code is sensible. Let's stay the course until we have performance data.

Thanks for looking into `JSBI.__compare()`!